### PR TITLE
Add arg: filename for screenshot

### PR DESCRIPTION
**Pull request by jamesgkk**
Add arg: filename to let user define the screenshot filename.

---
This change is [Reviewable](https://reviewable.io/reviews/google/mobly/730)

### CODE REVIEW

**Review by eric100lin — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/730)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @jamesgkk)

___
*[mobly/controllers/android_device.py, line 1008 at r1](https://reviewable.io/reviews/google/mobly/730#-MVFF99p4izt9Do_tu9c:-MVFF99p4izt9Do_tu9d:btykjgu) ([raw file](https://github.com/google/mobly/blob/a2d59e52531ecd96f4d3164f80706515b70d22c7/mobly/controllers/android_device.py#L1008)):*
> ```
> filename
> ```
This arg is `file_type` to `generate_filename` function.
Not the final file name of the screenshot.

**Comment by jamesgkk** [comment]
Add an arg to add additional name to the file.

**Review by jamesgkk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/730)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @eric100lin)

___
*[mobly/controllers/android_device.py, line 1008 at r1](https://reviewable.io/reviews/google/mobly/730#-MVFF99p4izt9Do_tu9c:-MVFSR234HbeWrVxOaWn:b44x7no) ([raw file](https://github.com/google/mobly/blob/a2d59e52531ecd96f4d3164f80706515b70d22c7/mobly/controllers/android_device.py#L1008)):*
<details><summary><i>Previously, eric100lin (Eric Lin (Tzu Hsiang Lin)) wrote…</i></summary><blockquote>

> ```
> filename
> ```
This arg is `file_type` to `generate_filename` function.
Not the final file name of the screenshot.
</blockquote></details>

Add a arg:filename to generate_filename for additional name of file.
It will only be added if given.

File: mobly/controllers/android_device.py [hunk 1]
```diff
@@ -897,19 +897,22 @@ def unload_snippet(self, name):
 
   def generate_filename(self,
                         file_type,
+                        filename=None,
```
Comment (xpconanfan) [inline mobly/controllers/android_device.py:900]: I don't think you actually need to add this?
you can simply use `file_type` to do what you need?

File: mobly/controllers/android_device.py [hunk 2]
```diff
@@ -995,16 +1000,17 @@ def take_bug_report(self,
     self.log.debug('Bugreport taken at %s.', full_out_path)
     return full_out_path
 
-  def take_screenshot(self, destination):
+  def take_screenshot(self, destination, filename=None):
     """Takes a screenshot of the device.
```
Comment (xpconanfan) [inline mobly/controllers/android_device.py:1005]: explain in docstring the detailed behavior of the `filename`
what happens if no filename is given etc.

**Review by xpconanfan — CHANGES_REQUESTED** [review]

File: mobly/controllers/android_device.py (hunk 1 above)
Comment (jamesgkk) [reply mobly/controllers/android_device.py:900]: May user confuse for the filename showing on the file_type?

Comment (xpconanfan) [reply mobly/controllers/android_device.py:900]: The "user" here is `AndroidDevice`'s own `take_screenshot`, so I don't think we should worry about it?

Comment (eric100lin) [reply mobly/controllers/android_device.py:900]: According to [go/mobly-ad-output-files](http://go/mobly-ad-output-files),
the file type naming convention is designed to make accessing generated files easily.
Please consider to change 'debug_tag' or override 'generate_filename' in your test class to have special, fully customized file name.

Comment (jamesgkk) [reply mobly/controllers/android_device.py]: Give the filename to screenshot directly, not modify the 'generate_filename' funtion

File: mobly/controllers/android_device.py [hunk 3]
```diff
@@ -995,16 +995,21 @@ def take_bug_report(self,
...
 
     Args:
       destination: string, full path to the directory to save in.
+      filename: string, additional name of the screenshot will be added after the
+        file_type:screenshot if given.
```
Comment (xpconanfan) [inline mobly/controllers/android_device.py:1004]: should not refer to `file_type` as it is not relevant to users using this method?

File: mobly/controllers/android_device.py [hunk 4]
```diff
@@ -995,16 +995,21 @@ def take_bug_report(self,
...
 
     Returns:
       string, full path to the screenshot file on the host.
     """
-    filename = self.generate_filename('screenshot', extension_name='png')
+    prefix = 'screenshot'
+    if filename:
+        prefix = 'screenshot_%s' % filename
```
Comment (xpconanfan) [inline mobly/controllers/android_device.py:1011]: use `f` string formatting

**Review by xpconanfan — CHANGES_REQUESTED** [review]

**Review by xpconanfan — CHANGES_REQUESTED** [review]
Need to add unit tests for the new code path

File: mobly/controllers/android_device.py [hunk 5]
```diff
@@ -1000,15 +1000,14 @@ def take_screenshot(self, destination, filename=None):
 
     Args:
       destination: string, full path to the directory to save in.
-      filename: string, additional name of the screenshot will be added after the
-        file_type:screenshot if given.
+      filename: string, additional name of the screenshot will be added if given.
```
Comment (xpconanfan) [inline mobly/controllers/android_device.py:1003]: How about we simply allow users to pass in `prefix`.
And the default value of `prefix` is "screenshot"?

**Review by xpconanfan — CHANGES_REQUESTED** [review]

File: mobly/controllers/android_device.py [hunk 6]
```diff
@@ -995,16 +995,18 @@ def take_bug_report(self,
     self.log.debug('Bugreport taken at %s.', full_out_path)
     return full_out_path
 
-  def take_screenshot(self, destination):
+  def take_screenshot(self, destination, filename='screenshot'):
     """Takes a screenshot of the device.
 
     Args:
       destination: string, full path to the directory to save in.
+      filename: string, prefix file name of the screenshot.
```
Comment (xpconanfan) [inline mobly/controllers/android_device.py:1003]: Name the arg `prefix`?

**Review by xpconanfan — CHANGES_REQUESTED** [review]

**Review by xpconanfan — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r8.
> *[Reviewable](https://reviewable.io/reviews/google/mobly/730#-:-M_-jY2R0lGRzDQw8OpD:bnfp4nl)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @eric100lin)

**Review by xpconanfan — DISMISSED** [review]

**Review by xpconanfan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/730)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @eric100lin and @xpconanfan)

___
*[mobly/controllers/android_device.py, line 1008 at r1](https://reviewable.io/reviews/google/mobly/730#-MVFF99p4izt9Do_tu9c:-M_-mRHG35e24iAOGACw:b-ch18lr) ([raw file](https://github.com/google/mobly/blob/a2d59e52531ecd96f4d3164f80706515b70d22c7/mobly/controllers/android_device.py#L1008)):*
<details><summary><i>Previously, jamesgkk wrote…</i></summary><blockquote>

Add a arg:filename to generate_filename for additional name of file.
It will only be added if given.

</blockquote></details>

This is ok.
Before we were using "screenshot" as the type, now we just allow users to override that with a string that's more contextual to their test case.

**Review by eric100lin — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/730)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @jamesgkk and @xpconanfan)

___
*[tests/mobly/controllers/android_device_test.py, line 854 at r9](https://reviewable.io/reviews/google/mobly/730#-M_-xQyF-PTuNSXffuNh:-M_-xQyF-PTuNSXffuNi:b-wk6fx3) ([raw file](https://github.com/google/mobly/blob/d7c859e74b616f9238ff05c26035e8387c6b043d/tests/mobly/controllers/android_device_test.py#L854)):*
> ```
> ··
> ```
remove extra spaces?
___
*[tests/mobly/controllers/android_device_test.py, line 868 at r9](https://reviewable.io/reviews/google/mobly/730#-M_-vxis-7gjIcHLiu2j:-M_-vxis-7gjIcHLiu2k:b-llg3iu) ([raw file](https://github.com/google/mobly/blob/d7c859e74b616f9238ff05c26035e8387c6b043d/tests/mobly/controllers/android_device_test.py#L868)):*
> ```
> screenshot_prefix
> ```
nit: better to just write the 'page_a' here instead of having a variable.
same issue with `mock_serial`
___
*[tests/mobly/controllers/android_device_test.py, line 868 at r9](https://reviewable.io/reviews/google/mobly/730#-M_-wPPk-HWfJsakJXCT:-M_-wPPk-HWfJsakJXCU:bm0a0n1) ([raw file](https://github.com/google/mobly/blob/d7c859e74b616f9238ff05c26035e8387c6b043d/tests/mobly/controllers/android_device_test.py#L868)):*
> ```
> full_pic_path = ad.take_screenshot(self.tmp_dir, screenshot_prefix)
> ```
Better to add empty lines before and after this line to make the test follow AAA rule.
go/unit-testing-practices#structure

**Review by eric100lin — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/google/mobly/730#-:-M_01Ctr6kUf0Q6PLwed:bnfp4nl)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @jamesgkk and @xpconanfan)

**Pull request merged**